modules/core/src/main/java/org/apache/ignite/internal/processors/cache/IgniteCacheOffheapManagerImpl.java

```diff
@@ -1280,22 +1239,21 @@ private long allocateForTree() throws IgniteCheckedException {
 
     /** {@inheritDoc} */
     @Override public final CacheDataStore createCacheDataStore(int p) throws IgniteCheckedException {
-        CacheDataStore dataStore;
-
         partStoreLock.lock(p);
 
         try {
-            assert !partDataStores.containsKey(p);
+            CacheDataStore store = createCacheDataStore0(p);
 
-            dataStore = createCacheDataStore0(p);
+            // Inject row cache cleaner on store creation.
+            // Used in case the cache with enabled SqlOnheapCache is single cache at the cache group.
+            if (ctx.kernalContext().query().moduleEnabled() && !grp.isLocal())
+                store.setRowCacheCleaner(ctx.kernalContext().indexProcessor().rowCacheCleaner(grp.groupId()));
```

Comment:
Why this part was moved from `GridDhtLocalPartition`? I see the only difference: now row cache cleaner is set under the lock, but looks like this lock doesn't protect anything related to row cache cleaner.

Reply:
Fixed.